### PR TITLE
feat: add Argo CD deployment investigation integration

### LINKED ISSUE #320: [FEATURE] Add ArgoCD integration for deployment and drift investigation
## Problem statement
OpenSRE currently lacks a first-class ArgoCD integration. That makes it harder to investigate rollout failures, configuration drift, sync issues, and deployment health in GitOps-driven Kubernetes environments.

## Proposed solution
Add a first-class ArgoCD integration that supports:
- onboarding, authentication, and connectivity verification
- reading application status, sync state, health, revision history, and recent changes
- using ArgoCD evidence during investigation planning and root-cause analysis
- surfacing deployment drift and recent rollout context clearly in reports

## Acceptance criteria
- the integration is usable from the investigation workflow end to end
- setup and usage are documented clearly
- the implementation is backed by end-to-end tests in `tests/e2e` or equivalent
- the PR includes a screen video showing setup plus a successful investigation flow using ArgoCD
- vibecoding without testing will not be allowed; PRs without working end-to-end coverage and screen-video proof should be rejected

### PR DESCRIPTION
**Pull request by MestreY0d4-Uninter**
Fixes #320

#### Describe changes -

This PR adds a first-class, read-only Argo CD integration for deployment and drift investigation.

What changed:
- Added a dedicated `app/services/argocd` client for application listing, application status/summary, managed resources, and server-side diff retrieval.
- Added Argo CD integration config, catalog/env resolution, and local verification support.
- Added read-only investigation tools:
  - `ArgoCDApplicationStatusTool`
  - `ArgoCDApplicationDiffTool`
- Wired Argo CD into source detection, evidence typing, investigation post-processing, and RCA evidence summaries.
- Added regression coverage for client behavior, catalog/verify resolution, source detection, and tools.

Security and safety hardening:
- Remote Argo CD URLs must use `https://`; plain `http://` is accepted only for localhost/loopback development endpoints, including IPv6 loopback.
- Bearer token auth and username/password session auth are supported, but ambiguous dual-auth configuration is rejected.
- Credentials, session tokens, bearer/JWT-like values, Kubernetes Secret diffs, private-key-like content, and password/token-like fields are redacted before surfacing errors, diffs, tool output, or evidence.
- Argo CD API usage remains read-only except for `POST /api/v1/session`, which is used only to obtain a session token when username/password auth is configured.
- HTTP 401 handling avoids unsafe retry loops and redacts sensitive response content.

### Screenshots (if UI) -

N/A — backend integration and investigation tooling only.

### Demo video -

https://github.com/MestreY0d4-Uninter/opensre/releases/download/pr-948-argocd-real-demo/opensre-pr948-argocd-real-walkthrough.mp4

Release artifact page: https://github.com/MestreY0d4-Uninter/opensre/releases/tag/pr-948-argocd-real-demo

What it shows:
- A real local Argo CD instance running on kind for the PR branch validation.
- The documented CLI path: `opensre integrations verify argocd`.
- Read-only Argo CD application status and list-mode evidence collection.
- A live GitOps drift scenario: the demo app is synced, the Kubernetes Deployment is scaled outside Git, Argo CD reports `OutOfSync`, and `ArgoCDApplicationDiffTool` returns `drift_detected=true` with a replica diff.
- Credentials are hidden/redacted; no raw token, password, JWT, or Kubernetes Secret value is shown.

Transparency note: this is a real terminal-session recording with commands entered into a live shell by automation for repeatability. The commands run against a local kind-hosted Argo CD instance and exercise the real OpenSRE code paths.

---

## Code Understanding and AI Usage

**Used AI?**
- [ ] No
- [x] Yes → confirmed:
  - [x] I reviewed every line of the AI-generated code
  - [x] I can explain the logic
  - [x] I tested edge cases
  - [x] I adapted it to the project's conventions

**Approach:**
- The implementation follows existing OpenSRE integration patterns: `app/services/*/client.py`, integration catalog/verify wiring, `BaseTool` investigation tools, source detection, and evidence/post-processing hooks.
- The first slice is intentionally read-only to keep the blast radius bounded while still supporting the issue's deployment/drift investigation use case.
- The main security focus was avoiding accidental credential exposure through configurable URLs, HTTP errors, Argo CD diffs, Kubernetes Secret manifests, and tool/evidence output.
- Edge cases covered include invalid/remote plaintext URLs, localhost/IPv6 loopback URLs, dual-auth rejection, whitespace-only list filters, auth failures, session-token 401 behavior, diff truncation, and env isolation.

---

## Validation

Follow-up validation after maintainer review:

- Addressed the `.env.example` request by adding the documented Argo CD environment variables.
- Verified the documented command against a real local Argo CD instance:
  - `uv run opensre integrations verify argocd`
  - Result: `argocd    local env    passed      Connected to Argo CD and listed 1 application.`
- Verified the real read-only status/list/diff tool paths against the same Argo CD instance.
- Verified drift detection with live Kubernetes drift:
  - `sync_status: OutOfSync`
  - `drift_detected: true`
  - `diff_count: 1`
  - replica drift shown for the `guestbook-ui` Deployment.
- Verified the negative path after restoring desired replicas:
  - `drift_detected: false`
  - `diff_count: 0`
- Focused regression suite after follow-up changes:
  - `245 passed`
- Quality gates after follow-up changes:
  - `make lint`: passed
  - `make format-check`: passed
  - `make typecheck`: passed
  - `git diff --check`: passed
- Demo video validation:
  - Duration: `42.93s`
  - Resolution: `1280x720`
  - SHA-256: `515c60b1e8a25c12302f5a6f48a5fb5ea697ed501b2c4ab8140ca898c2c9c5a7`
  - OCR/secret scan found no raw Argo CD password, bearer token, JWT, GitHub token, AWS key, or private key.

Local validation completed before opening the PR:

- Focused Argo CD tests:
  - `uv run pytest -q tests/services/argocd/test_client.py tests/integrations/test_argocd_catalog_verify.py tests/nodes/plan_actions/test_detect_sources_argocd.py tests/tools/test_argocd_tools.py`
  - Result: `33 passed`

- Expanded related tests:
  - `uv run pytest -q tests/integrations/test_argocd_catalog_verify.py tests/integrations/test_catalog_multi_instance.py tests/integrations/test_verify.py tests/nodes/plan_actions/test_detect_sources_argocd.py tests/tools/test_argocd_tools.py tests/tools/test_registry.py`
  - Result: `60 passed`

- Parallel focused run:
  - Argo CD focused suite under xdist with 4 workers
  - Result: `33 passed`

- Type checking:
  - `make typecheck`
  - Result: `Success: no issues found in 396 source files`

- Lint:
  - `make lint`
  - Result: `All checks passed!`

- Formatting:
  - `make format-check`
  - Result: `910 files already formatted`

- Whitespace/diff validation:
  - `git diff --check --cached`
  - Result: passed

- Coverage/full non-synthetic local suite:
  - `make test-cov`
  - Result: `3294 passed, 2 skipped, 1 xfailed, 1 warning`

- Local Argo CD scenario simulator:
  - Simulated Argo CD endpoints with a local `ThreadingHTTPServer`
  - Validated remote HTTP rejection, loopback HTTP allowance, dual-auth rejection, bearer list/status/diff flow, diff redaction, HTTP error redaction, lazy session login, verify surface, tools, and env catalog resolution.

---

## Checklist

- [x] Linked issue
- [x] Self-review completed
- [x] Purpose of each new function/class is clear from names, structure, and docstrings/comments where useful
- [x] Tests added for edge cases and regressions
- [x] Code style validated

### CODE REVIEW

**Comment by MestreY0d4-Uninter** [comment]
Updated in `b9e297f009f21cc112ff43d65167e8ff946f83bf` and addressed the Greptile findings:

- preserved the Argo CD `applications` list in list-mode evidence mapping instead of dropping it;
- narrowed bearer-prefix stripping so it only applies to bearer tokens, not generic string fields;
- added/updated regression coverage for both cases.

Validation:
- `uv run pytest -q tests/services/argocd/test_client.py tests/integrations/test_argocd_catalog_verify.py tests/nodes/plan_actions/test_detect_sources_argocd.py tests/tools/test_argocd_tools.py` — passed
- `make typecheck` — passed
- `make lint` — passed
- `make format-check` — passed
- `git diff --check` — passed
- GitHub checks on the updated head are green: quality, Analyze (python), typecheck, test, and CodeQL passed.

**Review by muddlebee — COMMENTED** [review]
looks good overall, few suggestions..

File: app/nodes/plan_actions/detect_sources.py [hunk 1]
```diff
@@ -1216,6 +1216,60 @@ def detect_sources(
...
+                raw_alert.get("error_message", ""),
+                annotations.get("summary", ""),
+                annotations.get("description", ""),
+                annotations.get("message", ""),
+            )
+            if value
+        ).lower()
+        has_gitops_hint = any(
+            marker in argocd_hint_text
+            for marker in ("argocd", "argo cd", "gitops", "drift", "deploy")
```
Comment (muddlebee) [inline app/nodes/plan_actions/detect_sources.py]: "deploy" is a substring match — "EKS pod deployment failed", "deploy pipeline", etc. all trigger this, adding Argo CD as a source for unrelated alerts. Replace with tighter Argo CD-specific terms: `("argocd", "argo cd", "argo-cd", "gitops", "outofsynced")`.

File: app/services/argocd/client.py [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+        return token
+
+    def _request(self, method: str, path: str, **kwargs: Any) -> httpx.Response:
+        token = self._ensure_session_token()
+        headers = {"Content-Type": "application/json"}
+        if token:
+            headers["Authorization"] = f"Bearer {token}"
+        headers.update(kwargs.pop("headers", {}) or {})
+        response = self._get_client().request(method, path, headers=headers, **kwargs)
+        response.raise_for_status()
```
Comment (muddlebee) [inline app/services/argocd/client.py]: If `raise_for_status()` throws a 401, `_session_token` is never cleared. A reused client instance will replay the expired token instead of re-authenticating. Clear `self._session_token = ""` when catching a 401 here, or document on the class that instances must not be reused after session expiry.

File: app/services/argocd/client.py [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+    r"|authorization\s*[:=]\s*\S+"
+    r"|xox[baprs]-[A-Za-z0-9-]{8,}"
+    r"|gh[pousr]_[A-Za-z0-9_]{20,}"
+    r"|AKIA[0-9A-Z]{16}"
+    r"|-----BEGIN [A-Z ]*PRIVATE KEY-----"
+    r"|eyJ[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,})"
+)
+
+
+def _is_loopback_host(host: str) -> bool:
```
Comment (muddlebee) [inline app/services/argocd/client.py]: _is_loopback_host and _validate_base_url are identical to the versions in app/integrations/models.py. Extract to app/utils/ or import from models to avoid the logic diverging.

File: app/nodes/root_cause_diagnosis/evidence_checker.py [hunk 4]
```diff
@@ -35,6 +35,7 @@
         "eks_node_health",
         "eks_pod_logs",
         "eks_deployment_status",
+        "argocd_application",
```
Comment (muddlebee) [inline app/nodes/root_cause_diagnosis/evidence_checker.py:39]: Only the singular "argocd_application" key is registered, so list-mode status calls (no application_name) never count toward evidence. If that is intentional, add a comment — the tool use_cases advertises listing as a valid scenario.

**Comment by MestreY0d4-Uninter** [comment]
Updated in `99ff09c55e5c2b3c70e205fccc6589d18ec78cbf` and addressed the latest Argo CD review comments:

- tightened Argo CD source hints by removing the generic `deploy`/`drift` substring triggers and keeping Argo CD/GitOps-specific markers;
- clear cached username/password session tokens on `401`, retry once with a fresh session, and keep retired tokens in the redaction set for surfaced errors;
- moved the HTTPS-or-loopback HTTP URL validation into a shared `app.utils.url_validation` helper used by both integration config and the service client;
- registered `argocd_applications` as investigated evidence so list-mode status calls count toward the evidence gate.

Validation run locally:
- `uv run pytest -q tests/services/argocd/test_client.py tests/integrations/test_argocd_catalog_verify.py tests/nodes/plan_actions/test_detect_sources_argocd.py tests/tools/test_argocd_tools.py tests/nodes/root_cause_diagnosis/test_healthy_short_circuit.py -q` → 205 passed
- `make lint`
- `make format-check`
- `make typecheck`
- `git diff --check`

**Comment by muddlebee** [comment]
Nice work. Demo video is missing though? Can you quickly add in the PR description

**Comment by MestreY0d4-Uninter** [comment]
Follow-up self-review fix in `dce3da0bc2d23e188e641d0031eeb951c6ee33b8`:

- added `argocd_diff` to investigated evidence so diff-only Argo CD investigations count the same way status/list-mode investigations do;
- included Argo CD status/list/diff keys in evidence availability so diagnosis does not treat Argo CD-only evidence as missing;
- added regression coverage for the healthy short-circuit and evidence-availability paths.

Additional validation:
- `uv run pytest -q tests/nodes/root_cause_diagnosis/test_evidence_checker.py tests/nodes/root_cause_diagnosis/test_healthy_short_circuit.py tests/tools/test_argocd_tools.py tests/services/argocd/test_client.py tests/integrations/test_argocd_catalog_verify.py tests/nodes/plan_actions/test_detect_sources_argocd.py -q` → 232 passed
- `make lint`
- `make format-check`
- `make typecheck`
- `git diff --check`

**Comment by MestreY0d4-Uninter** [comment]
Added the demo video to the PR description.

It shows the Argo CD setup/configuration, local verification, read-only status/list/diff paths, and secret/diff redaction behavior. Transparency note: the recording uses a live shell with automated input and a local Argo CD-compatible endpoint, so the real OpenSRE code path is exercised without exposing live credentials or Kubernetes Secrets.

**Comment by muddlebee** [comment]
> setup and usage are documented clearly

missing as per **Acceptance criteria**

https://github.com/Tracer-Cloud/opensre/issues/320

**Comment by MestreY0d4-Uninter** [comment]
Thanks for the catch. Addressed in the latest commit (`250c02a`).

Added `docs/argocd.mdx` and registered it in the docs navigation. The page now covers:

- setup via environment variables, persistent store, and `ARGOCD_INSTANCES`
- token vs username/password auth, including the no-dual-auth rule
- HTTPS/loopback HTTP behavior and TLS verification
- `opensre integrations verify argocd`
- investigation usage, supported alert hints, and the two Argo CD evidence tools
- read-only scope and secret/diff redaction behavior

I also updated the multi-instance integrations page to include `ARGOCD_INSTANCES`.

Local validation passed:
- `python -m json.tool docs/docs.json`
- `git diff --check`
- Mintlify local preview served `/argocd` successfully
- `make lint`
- `make format-check`
- `make typecheck`
- focused Argo CD/evidence pytest slice: 65 passed

Update: the first GitHub Actions run was cancelled before a hosted runner acquired the jobs. I re-pushed the same tree to trigger fresh checks; current head `250c02a` is now green (`quality`, `typecheck`, `test`, `Analyze (python)`, and CodeQL passed; Kubernetes/thorough jobs were skipped by workflow rules).

**Review by muddlebee — DISMISSED** [review]

**Comment by muddlebee** [comment]
@MestreY0d4-Uninter you still need to add entry here https://github.com/Tracer-Cloud/opensre/blob/main/.env.example for the env vars you created and document it!!

and in your demo I dont understand the command you ran? could you pls run any of the scenario mentioned in the docs instead? and not some unit tests pls

**Comment by muddlebee** [comment]
and pls give a walk through whats happening in demo

File: app/integrations/catalog.py [hunk 5]
```diff
@@ -84,6 +85,9 @@
     "opensearch": "opensearch",
     "open search": "opensearch",
     "alertmanager": "alertmanager",
+    "argocd": "argocd",
+    "argo cd": "argocd",
+    "argo-cd": "argocd",
```
Comment (VaibhavUpreti) [inline app/integrations/catalog.py]: do we need all of them? probably best to just stick to one

**Comment by MestreY0d4-Uninter** [comment]
Thanks for the follow-up notes. I addressed the remaining items in the latest head commit (`70e8c32`):

- added the Argo CD env vars to `.env.example`;
- made the documented `opensre integrations verify argocd` path work from the main CLI;
- kept the catalog key canonical as `argocd` only, while preserving Argo CD spelling variants only for alert/source hints;
- replaced the PR description demo with a real local kind-hosted Argo CD walkthrough instead of the earlier compatible-endpoint demo;
- verified read-only status/list/diff against that real Argo CD instance, including a live Deployment replica drift scenario where Argo CD reports `OutOfSync` and `ArgoCDApplicationDiffTool` returns `drift_detected=true`.

Validation after the follow-up changes:

- focused related pytest slice: `245 passed`;
- `make lint`: passed;
- `make format-check`: passed;
- `make typecheck`: passed;
- `git diff --check`: passed;
- GitHub Actions on the current head are green (`quality`, `typecheck`, `test`, `Analyze`/CodeQL; Kubernetes/thorough jobs are skipped by the workflow).

The PR description now includes the updated demo video link and a walkthrough of what is happening in the demo. Credentials are hidden/redacted, and the video OCR/secret scan did not find raw passwords, bearer tokens, JWTs, GitHub tokens, AWS keys, or private keys.

**Review by muddlebee — APPROVED** [review]
LGTM

**Review by yashksaini-coder — COMMENTED** [review]
## Senior Review — PR #948: Argo CD Deployment Investigation Integration

### Overall Assessment
**Solid, production-quality implementation. One hard blocker (merge conflicts), the rest is minor.** The contributor has been extremely responsive — Greptile's P1 (list-mode evidence loss), all three muddlebee P2 issues, and the `"deploy"` keyword broadness were all addressed in follow-up commits. Architecture follows established patterns cleanly.

---

### ✅ What's Done Well

**Security hardening (thorough)**
- `HTTPS`-only with loopback HTTP exception, extracted to `app/utils/url_validation` (shared with models) — no logic duplication
- Dual-auth rejection in `ArgoCDConfig._no_dual_auth()`
- Bearer token stripped from non-token field validators correctly (fixed from Greptile's original P2)
- Credentials, session tokens, and retired tokens all included in `_redact()` — retired token set is a nice touch
- 401 session retry with `_retired_session_tokens` clearing — prevents stale token replay
- `_SECRET_LINE_RE` + `_GENERIC_SECRET_VALUE_RE` diff sanitization covers JWTs, GitHub tokens, AWS keys, private keys

**Architecture follows project conventions**
- `ArgoCDConfig(StrictConfigModel)` ✅ (unlike the original Airflow PR which used `BaseModel`)
- `source="argocd"` ✅ (correct for prioritizer scoring)
- `"argocd"` in `EvidenceSource` Literal ✅
- `argocd_applications` + `argocd_diff` registered in evidence gate ✅
- URL validation in `app/utils/url_validation` (shared helper, not duplicated) ✅
- Docs in `docs/argocd.mdx` with nav registration ✅
- `.env.example` updated ✅
- Multi-instance support via `ARGOCD_INSTANCES` ✅

**Greptile P1 fixed**
- `_map_argocd_application_status` now correctly reads `data.get("applications", [])` (plural), so list-mode evidence is preserved

**Hint keywords tightened (muddlebee P2 fixed)**
- `"deploy"`, `"drift"` removed; now only `"argocd"`, `"argo cd"`, `"argo-cd"`, `"gitops"`, `"outofsync"`, `"outofsynced"`

**Tests: 44 pass, comprehensive coverage**
- Client auth flows (bearer, session login, dual-auth rejection, 401 retry)
- Diff sanitization and truncation
- Catalog/verify resolution, multi-instance
- Source detection with tightened hint markers
- Tool behavior (status single-app, list-mode, diff, not-configured path)

---

### 🔴 Blocker — Merge Conflicts

`mergeable_state: dirty` — `main` has moved significantly (201 files diverged) since this PR was last rebased. The PR **cannot be merged until conflicts are resolved**.

Key files that will need attention on rebase:
- `app/integrations/catalog.py` (Airflow integration was merged into main via #570)
- `app/nodes/investigate/processing/post_process.py`
- `app/nodes/plan_actions/detect_sources.py`
- `app/types/evidence.py` (Airflow added `"airflow"` to `EvidenceSource`)
- `.env.example`

**Action needed:** Rebase `issue/320-argocd-integration` onto latest `main`, resolve conflicts, and push. All quality gates should re-pass since the Argo CD changes are orthogonal to what landed in main.

---

### ⚠️ Minor Observations (non-blocking)

**`VaibhavUpreti`'s catalog.py comment** — "do we need all of them?" Looking at the code, `ARGOCD_AUTH_TOKEN` with `ARGOCD_TOKEN` as fallback is intentional to handle both naming conventions (some Argo CD tooling exports `ARGOCD_TOKEN`). This is fine and follows the same fallback pattern used elsewhere.

**`surfaces` not set on either tool** — `ArgoCDApplicationStatusTool` and `ArgoCDApplicationDiffTool` default to investigation-only. Given these tools are read-only and investigation-focused, that's appropriate. Adding `surfaces=("investigation", "chat")` would be a cosmetic consistency improvement but is not blocking.

---

### Summary

| Category | Status |
|---|---|
| Code quality | ✅ Clean (ruff, mypy) |
| Tests | ✅ 44 pass |
| Security hardening | ✅ Thorough |
| Architecture fit | ✅ Follows conventions |
| Greptile P1 (list-mode) | ✅ Fixed |
| muddlebee P2 issues | ✅ Fixed |
| Docs + .env.example | ✅ Complete |
| CI checks (last head) | ✅ All green |
| **Merge conflicts** | ❌ **Blocker — rebase onto main required** |

Once the rebase is done and CI is green again, this is mergeable.

**Comment by muddlebee** [comment]
@MestreY0d4-Uninter pls fix merge conflicts

**Comment by MestreY0d4-Uninter** [comment]
Thanks @muddlebee, fixed.

I rebased the branch onto the latest `main` and force-pushed the refreshed PR head (`8b0ebb67b39fa9babc96c36c9622a2d6450ca799`). The merge-conflict blocker is resolved now: GitHub reports `mergeStateStatus: CLEAN` and the PR still has `reviewDecision: APPROVED`.

Validation:
- `python -m json.tool docs/docs.json`
- focused Argo CD/regression slice: 245 passed
- MoA-claim spot checks for Argo CD evidence/session/source-detection paths: 6 passed
- `make lint`
- `make format-check`
- `make typecheck`
- `make test-cov`: 4053 passed, 2 skipped, 1 xfailed
- GitHub checks are green: `quality`, `typecheck`, `test`, `Analyze (python)`, `CodeQL`

I also re-checked the earlier review items against the current head; no remaining blocker found.

**Comment by muddlebee** [comment]
@MestreY0d4-Uninter nice work. and from next time lets try to keep the commits after rebase/resolve conflicts clean :) difficult to spot any errors or missing..
